### PR TITLE
fix #14157

### LINKED ISSUE #14157: Setting subscript index of cstring is allowed and generates bad code
Low priority, disallowing this makes more sense than making it work.

### Example
```nim
var x = cstring"abcd"
x[0] = 'x'
```

### Current Output
```
x_43003[0].charCodeAt(0) = 97;
^

ReferenceError: Invalid left-hand side in assignment
```

### Expected Output
compile error

### Possible Solution

* undefine []= for cstring in JS. Or leave it alone and let people get the error on their own. relevant definition:

https://github.com/nim-lang/Nim/blob/dc3919bb1af89799e391b4c4ecd0f1f60f7862ff/lib/system.nim#L437-L438

```
$ nim -v
Nim Compiler Version 1.3.1
```

### PR DESCRIPTION
**Pull request by ringabout**

### CODE REVIEW

File: compiler/jsgen.nim [hunk 1]
```diff
@@ -1049,6 +1049,10 @@ proc genAsgnAux(p: PProc, x, y: PNode, noCopyNeeded: bool) =
   var a, b: TCompRes
   var xtyp = mapType(p, x.typ)
 
+  # disable `[]=` for cstring
+  if x.kind == nkBracketExpr and x.len >= 2 and x[0].typ.kind == tyCString:
```
Comment (Araq) [inline compiler/jsgen.nim]:
```suggestion
  if x.kind == nkBracketExpr and x.len >= 2 and x[0].typ.skipTypes(abstractInst).kind == tyCString:
```

**Comment by juancarlospaco** [comment]
Should it have a line on the changelog?. :thinking:

**Comment by timotheecour** [comment]
where is `[]=` even defined for `cstring` ? I commented out a few suspects but it doesn't give CT error (would seem cleaner to disable it by patching the relevant `[]=` proc, but not sure how to do that); https://github.com/nim-lang/Nim/issues/15911 also makes it hard (from `system import nil` doesn't help with disabling `[]=`).

I'm guessing `[]=` is treated specially for string/cstring in the compiler, but is there a reason to?

**Comment by ringabout** [comment]
I'm confused too. That's the reason why I choose to change JS gen.
https://github.com/nim-lang/Nim/issues/14157#issuecomment-723902349

File: compiler/jsgen.nim [hunk 2]
```diff
@@ -1049,6 +1049,10 @@ proc genAsgnAux(p: PProc, x, y: PNode, noCopyNeeded: bool) =
   var a, b: TCompRes
   var xtyp = mapType(p, x.typ)
 
+  # disable `[]=` for cstring
+  if x.kind == nkBracketExpr and x.len >= 2 and x[0].typ.skipTypes(abstractInst).kind == tyCString:
+    localError(p.config, x.info, "cstring doesn't support `[]=` operator!")
```
Comment (Araq) [inline compiler/jsgen.nim]:
```suggestion
    localError(p.config, x.info, "cstring doesn't support `[]=` operator")
```

File: tests/js/tmodify_cstring.nim [hunk 3]
```diff
@@ -0,0 +1,6 @@
+discard """
+  errormsg: "cstring doesn't support `[]=` operator!"
```
Comment (Araq) [inline tests/js/tmodify_cstring.nim]:
```suggestion
  errormsg: "cstring doesn't support `[]=` operator"
```

**Review by Araq — APPROVED** [review]

**Pull request merged**